### PR TITLE
DRY up tracking area handling

### PR DESCRIPTION
**Pull request by sindresorhus**

### CODE REVIEW

**Comment by sindresorhus** [comment]
@boyvanamstel Thoughts?

File: Sources/CircularProgress/CustomButton.swift [hunk 1]
```diff
@@ -2,6 +2,41 @@ import Cocoa
 
 // TODO(sindresorhus): I plan to extract this into a reusable package when it's more mature.
 
+final class TrackingArea {
+	private let view: NSView
+	private let rect: CGRect
```
Comment (boyvanamstel) [inline Sources/CircularProgress/CustomButton.swift:27]: As far as I can see [`NSTrackingArea`'s rect](https://developer.apple.com/documentation/appkit/nstrackingarea/1524488-init) and the [view's bounds](https://developer.apple.com/documentation/appkit/nsview/1483817-bounds) are `NSRect`.

**Review by boyvanamstel — COMMENTED** [review]
In practice I think it's unusual to have two separate implementations this close to each other, but this is rather nice in those cases. 👍

File: Sources/CircularProgress/CustomButton.swift (hunk 1 above)
Comment (sindresorhus) [reply Sources/CircularProgress/CustomButton.swift:27]: `NSRect` and `CGRect` are the same thing. I prefer the latter naming for consistency.

See: https://github.com/realm/SwiftLint/issues/2371

**Comment by sindresorhus** [comment]
> In practice I think it's unusual to have two separate implementations this close to each other

Can you elaborate? Not sure I understand what you mean by "it's unusual to have two separate implementations this close to each other".

The reason I made it a class is that I have the exact same boilerplate tracking setup and updating in a lot of internal code. I have also seen it in various open source projects. So I thought it would be nice to hide away the boilerplate.

File: Sources/CircularProgress/CustomButton.swift (hunk 1 above)
Comment (boyvanamstel) [reply Sources/CircularProgress/CustomButton.swift:27]: I don't think there's `NSTrackingArea` on iOS, but I see what you mean. :)

**Review by boyvanamstel — APPROVED** [review]

**Comment by boyvanamstel** [comment]
> Can you elaborate? Not sure I understand what you mean by "it's unusual to have two separate implementations this close to each other".

In my own experience I usually end up creating a subclass that handles hovering (which contains the tracking area code), or try to have just one view that registers mouse movement and which informs others. This means I don't often have more than one instance where I override `updateTrackingAreas()`.

That said, it's useful here.

**Comment by sindresorhus** [comment]
I try to subclass as little as possible. I prefer composition. Especially, since Swift does not support multiple inheritance, so you can't create a generic "HoverView" subclass and use it for `NSButton` too.

**Comment by sindresorhus** [comment]
Could maybe make a `AreaTrackable` protocol. But we would need to configure it somehow. Not sure how that would work in practice.

**Comment by boyvanamstel** [comment]
> I try to subclass as little as possible. I prefer composition. Especially, since Swift does not support multiple inheritance, so you can't create a generic "HoverView" subclass and use it for `NSButton` too.

Same here. Views (especially buttons) that hover are the exception, for me.

**Comment by boyvanamstel** [comment]
> Could maybe make a `AreaTrackable` protocol. But we would need to configure it somehow. Not sure how that would work in practice.

Something like this?

```swift
protocol AreaTrackable where Self: NSView {
  var trackingArea: NSTrackingArea { get }
  var trackingRect: CGRect { get }
  var trackingOptions: NSTrackingArea.Options { get }
  func update()
}
extension AreaTrackable {
  func update() {
    ...
  }
}
```

**Comment by boyvanamstel** [comment]
That would only allow for a single tracking area though. Not ideal. The `TrackingArea` class is a better option, I think.

File: Sources/CircularProgress/CustomButton.swift [hunk 2]
```diff
@@ -2,6 +2,41 @@ import Cocoa
...
+	private let options: NSTrackingArea.Options
+	private var trackingArea: NSTrackingArea?
+
+	/**
+	- Parameter rect: Defaults to `view.bounds`
+	*/
+	init(for view: NSView, rect: CGRect? = nil, options: NSTrackingArea.Options = []) {
```
Comment (sindresorhus) [inline Sources/CircularProgress/CustomButton.swift:36]: What do you think about renaming `rect` to `area`?

Comment (boyvanamstel) [reply Sources/CircularProgress/CustomButton.swift:36]: I like `rect`. It corresponds directly with the argument on `NSTrackingArea`'s init.

**Comment by sindresorhus** [comment]
I added some docs. Good to merge?

**Comment by boyvanamstel** [comment]
> I added some docs. Good to merge?

Yep. :)

**Comment by sindresorhus** [comment]
Thanks for the review :)

**Pull request merged**